app/scripts/lib/indexed-db-backend.ts

```diff
@@ -0,0 +1,185 @@
+type StorageKey = {
+  name: string;
+  chainId: string;
+};
+
+type StorageBackend = {
+  read(key: StorageKey, checksum: string): Promise<ArrayBuffer>;
+  write(key: StorageKey, data: ArrayBuffer, checksum: string): Promise<void>;
+  delete(key: StorageKey): Promise<void>;
+  dir(): Promise<StorageKey[]>;
+};
+
+const validateChecksum = async (
+  key: StorageKey,
+  data: ArrayBuffer,
+  checksum: string,
+) => {
+  const hash = await crypto.subtle.digest('SHA-256', data);
+  const hashString = Array.from(new Uint8Array(hash))
+    .map((b) => b.toString(16).padStart(2, '0'))
+    .join('');
+
+  if (hashString !== checksum) {
+    throw new Error(`Checksum mismatch for key ${key}`);
+  }
+};
+
+export class IndexedDBBackend implements StorageBackend {
```

Comment:
Should we be more explicit here and call this something like IndexexDBPPOMStorage as it exists to support that and is using its interface?